src/v/coproc/pacemaker_context.h

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2020 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "coproc/supervisor.h"
+#include "coproc/types.h"
+#include "rpc/reconnect_transport.h"
+#include "storage/api.h"
+#include "storage/log.h"
+#include "units.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/core/semaphore.hh>
+
+#include <absl/container/flat_hash_map.h>
+
+namespace coproc {
+/// Structure representing state about input topics that scripts will subscribe
+/// to. ss::shared_ptrs to ntp_contexts will be used as there may be many
+/// subscribers to an input ntp
+struct ntp_context {
+    explicit ntp_context(storage::log lg, script_id id)
+      : log(std::move(lg)) {
+        offsets.emplace(id, model::offset(0));
+    }
+
+    const model::ntp& ntp() const { return log.config().ntp(); }
+
+    /// Reference to the storage layer for reading from the input ntp
+    storage::log log;
+    /// Interested scripts write their last read offset of the input ntp
+    absl::flat_hash_map<script_id, model::offset> offsets;
```

Comment:
very memory hungry, make sure you push saturation of the container to like 100% we don't need amortization, we need exact use.